[build] Add Docker images with FIPS

Ikuni17 opened the pull request:
## Summary
Closes elastic/kibana-operations/issues/24

This adds a second flavor of UBI image (`kibana-ubi-fips`) which has a FIPS compliant version of OpenSSL compiled and linked to Node. Using the label `ci:build-docker-fips` will create the image in CI and push to the registry.

The FIPS image start the Kibana NodeJS process using the FIPS compliant OpenSSL version. Kibana will start in this state but crash during runtime because there are many code changes required for it to be FIPS compliant, including `node_module` usage. I attempted numerous ways to load other OpenSSL providers alongside the FIPS provider, but it always led to Kibana crashing on invalid algorithm usage.

Ikuni17 committed:
Base template changes from build repo

Ikuni17 committed:
Cleanup

Ikuni17 committed:
Add task and flags for fips build

Ikuni17 committed:
Add help text

Ikuni17 committed:
Fix snapshots

Ikuni17 committed:
add cli test

Ikuni17 committed:
Fix fips flag not passed all the way

jbudz commented on src/dev/build/tasks/os_packages/create_os_package_tasks.ts:
```diff
@@ -172,3 +172,26 @@ export const CreateDockerContexts: Task = {
+
+  async run(config, log, build) {
+    await runDockerGenerator(config, log, build, {
+      architecture: 'x64',
```
Mind if we also add ARM builds?  I expect we'll want both eventually, but mostly now so I can run a test build.

Ikuni17 replied:
I will try to circle back to this or make a new issue for it. I ran a build using `aarch64` but the OpenSSL compilation step was running for ~35 minutes before I cancelled it. From a bit of research, it looks like we will have to compile OpenSSL differently for ARM as well.

Ikuni17 committed:
Add arm image to build tasks

Ikuni17 committed:
Update docs

Ikuni17 committed:
Fix dockerfile whitespace

Ikuni17 committed:
Add comment to openssl cnf

Ikuni17 committed:
Handle build fips flag

Ikuni17 committed:
Revert ARM build

Ikuni17 committed:
set owner for openssl dir

Ikuni17 commented:
I added support to build FIPS images in CI with 18986b2aec65f8404b5c2335bd61d7f7af01d460

For now they are in the build distribution step artifacts: https://buildkite.com/elastic/kibana-pull-request/builds/190640#018d6276-87ab-4ef9-9d3e-17d249ac0e79

The images built in CI are not loading FIPS properly due to self test failing, so investigating.

Ikuni17 commented on .buildkite/scripts/build_kibana.sh:
```diff
@@ -22,27 +22,42 @@ node scripts/build "${BUILD_ARGS[@]}"
 EOF
 fi
 
+if is_pr_with_label "ci:build-fips-images"; then
```
I think we should move this and the cloud image build into a new step after this PR, so they do not delay CI from time increase on Build Kibana Distribution and Plugins step.

Ikuni17 committed:
test only download openssl

Ikuni17 commented:
/ci

Ikuni17 committed:
Revert "test only download openssl"

This reverts commit e8c403f2a12c7aa9e07308df9bf70d44da949aa7.

Ikuni17 committed:
Adjust cmd order. Split make step for debug

Ikuni17 committed:
Disable ubi

Ikuni17 committed:
Fix silent errors. Serial commands

Ikuni17 committed:
Fix shasum

Ikuni17 committed:
disable legacy openssl in fips mode

Ikuni17 committed:
Revert "Disable ubi"

This reverts commit 44b4e14361e2c8b85de353a82c460f73a2171ca0.

Ikuni17 committed:
Revert set owner for openssl dir

Ikuni17 committed:
override env var for os wide fips

Ikuni17 committed:
Revert "Revert set owner for openssl dir"

This reverts commit 6830fc7b0f98ff09cdb06fcade99ff2cacd65c08.

Ikuni17 committed:
Revert "override env var for os wide fips"

This reverts commit c3201ca1fa595d90a7a189ec5e23af1ff8cdf2a4.

Ikuni17 commented:
/ci

mistic committed:
fix(NA): correctly setup FIPS when using the Kibana docker generator

docs(NA): change wording on tutorials

chore(NA): correctly bundle docker contexts

chore(NA): copy fips related resources in fips only

fix(NA): dockerfile using -e on echo

fix(NA): remove extra char

Ikuni17 committed:
add set x for debug

Ikuni17 committed:
slight command reorder

Ikuni17 committed:
Remove skip context

Ikuni17 committed:
adjust build params

Ikuni17 committed:
Add fips pipeline

Ikuni17 committed:
Merge remote-tracking branch 'upstream/main' into feature/24/docker-img-fips-provider

Ikuni17 committed:
bypass checks for now

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

Ikuni17 committed:
Merge branch 'feature/24/docker-img-fips-provider' of github.com:Ikuni17/kibana into feature/24/docker-img-fips-provider

Ikuni17 committed:
Increase timeout

Ikuni17 committed:
Add artifact upload

mistic committed:
refact(NA): re-arrange fips generation with the rest of the build process

Ikuni17 committed:
cleanup extra whitespace changes

Ikuni17 committed:
add docker login

cauemarcondes reviewed (approved):
Profiling LGTM

Ikuni17 committed:
remove ubuntu build pkgs

Ikuni17 committed:
Rremove skip archives

Ikuni17 committed:
Use bk commit env var

Ikuni17 committed:
Merge remote-tracking branch 'upstream/main' into feature/24/docker-img-fips-provider

Ikuni17 commented:
/ci

Ikuni17 committed:
download artifact from build step

Ikuni17 committed:
fix plural

Ikuni17 committed:
add namespace

mistic committed:
Merge branch 'main' into feature/24/docker-img-fips-provider

mistic commented on .buildkite/scripts/steps/fips/build.sh:
```diff
@@ -0,0 +1,35 @@
+
+# Moving to `target/` first will keep `buildkite-agent` from including directories in the artifact name
+cd "$KIBANA_DIR/target"
+buildkite-agent artifact upload "./*.tar.gz;./*.zip;./*.deb;./*.rpm"
```
```suggestion
buildkite-agent artifact upload "./*docker-image*.tar.gz"
```

mistic replied:
Not critical but I think we only need to upload the docker images here

Ikuni17 replied:
Thanks, was thinking the same

Ikuni17 committed:
Only upload docker image artifacts

Co-authored-by: Tiago Costa <tiago.costa@elastic.co>

mistic reviewed (approved):
lgtm
